### DISCUSSION TITLE
How to only compile certain directories depending on target framework like in MAUI?

### DISCUSSION THREAD

**Discussion by KieranDevvs** [dotnet/msbuild#7782] [Q&A] [answered]
I'm trying to replicate the system that MAUI uses to separate platform / target dependant code without using the pre processor (`#if def`) or partial classes / methods.

In MAUI, the convention is that there is a platform directory that contains each build target and only the relevant code is compiled, thus effectively swapped out for each build.

I've checked the csproj file but there seems to be no reference to `platform` or any such directory / build steps.

Is this specifically a MAUI convention, or is there a way to replicate this behaviour in other project types?

![image](https://user-images.githubusercontent.com/21192520/176974434-56efabac-79a3-4b93-811e-7cc9fcf35e74.png)

Ive tried to remove the root directory and then re-add the the relevant sub folder depending on the target framework but roslyn complains about objects being defined multiple times (i.e its seeing files for all target frameworks).

```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFrameworks>net5;net6</TargetFrameworks>
        <Nullable>enable</Nullable>
        <LangVersion>latest</LangVersion>
    </PropertyGroup>

    <ItemGroup>
        <Compile Remove="Platforms\**" />
    </ItemGroup>
    
    <ItemGroup Condition=" '$(TargetFramework)' == 'net5' ">
        <Compile Include="Platforms\net5\**" />
    </ItemGroup>

    <ItemGroup Condition=" '$(TargetFramework)' == 'net6' ">
        <Compile Include="Platforms\net6\**" />
    </ItemGroup>

</Project>

```

Thanks.

**Comment by rainersigwald** [comment, marked as answer]
MAUI does this via some fairly complicated, general logic in one of its `.targets` files: https://github.com/dotnet/maui/blob/c1439acf4dfeb0e5a6f56583d6a1860a12ff4a6f/.nuspec/Microsoft.Maui.Controls.SingleProject.targets#L72-L119

You can accomplish similar things with a more straightforward `Compile Remove="myThing/**"`, though you should be aware of some behaviors in the project system that get confused with just `Remove`; I think that's why MAUI does the `ExcludeFromCurrentConfiguration` thing.
